lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -1489,3 +1496,158 @@ void ConstExprEvaluator::computeConstantValues(
     numInstEvaluated = 0;
   }
 }
+
+//===----------------------------------------------------------------------===//
+// ConstExprStepEvaluator implementation.
+//===----------------------------------------------------------------------===//
+
+ConstExprStepEvaluator::ConstExprStepEvaluator(SymbolicValueAllocator &alloc,
+                                               SILFunction *fun)
+    : evaluator(ConstExprEvaluator(alloc)),
+      internalState(
+          new ConstExprFunctionState(evaluator, fun, {}, stepsEvaluated)) {
+  assert(fun);
+}
+
+ConstExprStepEvaluator::~ConstExprStepEvaluator() { delete internalState; }
+
+Optional<SymbolicValue> ConstExprStepEvaluator::incrementStepsAndCheckLimit(
+    SILInstruction *inst, bool includeInInstructionLimit) {
+  if (includeInInstructionLimit && ++stepsEvaluated > ConstExprLimit) {
+    // Note that there is no call stack to associate with the unknown value
+    // as the interpreter is at the top-level here.
+    return SymbolicValue::getUnknown(inst, UnknownReason::TooManyInstructions,
+                                     {}, evaluator.getAllocator());
+  }
+  return None;
+}
+
+std::pair<Optional<SILBasicBlock::iterator>, Optional<SymbolicValue>>
+ConstExprStepEvaluator::evaluate(SILBasicBlock::iterator instI,
+                                 bool includeInInstructionLimit) {
+  // Diagnose whether evaluating this instruction exceeds the instruction
+  // limit, unless asked to not include this instruction in the limit.
+  auto limitError =
+      incrementStepsAndCheckLimit(&(*instI), includeInInstructionLimit);
+  if (limitError) {
+    return {None, limitError.getValue()};
+  }
+  return internalState->evaluateInstructionAndGetNext(instI, visitedBlocks);
+}
+
+std::pair<Optional<SILBasicBlock::iterator>, Optional<SymbolicValue>>
+ConstExprStepEvaluator::skip(SILBasicBlock::iterator instI,
+                             bool includeInInstructionLimit) {
+  SILInstruction *inst = &(*instI);
+
+  // Diagnose whether evaluating this instruction exceeds the instruction
+  // limit, unless asked to not include this instruction in the limit.
+  auto limitError =
+      incrementStepsAndCheckLimit(inst, includeInInstructionLimit);
+  if (limitError) {
+    return {None, limitError.getValue()};
+  }
+
+  // Reset all constant state that could be mutated by the instruction
+  // to an unknown symbolic value.
+  for (auto &operand : inst->getAllOperands()) {
+    auto constValOpt = lookupConstValue(operand.get());
+    if (!constValOpt) {
+      continue;
+    }
+    auto constVal = constValOpt.getValue();
+    if (constVal.getKind() != SymbolicValue::Address) {
```

Comment:
Looking at this again after having just thought about https://github.com/apple/swift/pull/22772, I realize that this will have to be careful about aggregates containing addresses (and also addresses pointing at aggregates containing addresses).

There is no such thing as an aggregate containing an address now. Even after arrays are merged, this won't have to handle the array aggregates containing an address, because they have value semantics. (I haven't thought this through super carefully yet, but I think it's true.)

So I'm reasonably sure that no action is required now.

But if/when the constant evaluator gets reference types, this will be a problem. (Perhaps it will never get reference types during top level evaluation, exactly because of problems like this.) Is there a way to make sure we don't forget?

Reply:
Actually, I remembered that reference types show up as reference types in SIL, not as addresses. So there should never be any aggregates containing addresses (except for things like arrays that are value types but that contain addresses as an implementation detail). So this will not have to worry about addresses in aggregates.

The point that it will have to worry about reference types is still valid. But we don't have reference types now.

Reply:
Those are interesting observations. You are right that if the interpreter starts supporting reference types (or structs with reference types), this way of estimating effects is not right. We need to do a traversal of the reachable memory locations to correctly approximate all possible effects.

As you say Arrays and Strings are fine as they have value semantics. It does seem to me that we don't need any additional work for Arrays. 

It doesn't look like "Structs with addresses"  is a thing in SIL, unless we allow unsafe pointers, which can store addresses. However, allowing anything like that in the interpreted fragment does seem messy.  

> Is there a way to make sure we don't forget?

I think we can document code, add a test that checks that diagnostics are emitting for skip with reference types, and also add an assert to `skip` that will fail on any new kind of symbolic object that is not currently supported. This will force anyone who is adding a new kind of symbolic object to look at `skip`.